### ISSUE TITLE
Incorrect Parameter Name: "d_coeff" Should Be "d_coef"

### ISSUE THREAD

**Issue by Packsod** [kijai/ComfyUI-FluxTrainer#149]
![Image](https://github.com/user-attachments/assets/77d942a0-f960-47aa-9dda-30a027639e1b)

**Description:**

There is a mistake in the parameter naming for the **"Optimizer Config ProdigyPlusScheduleFree"** node. The parameter **"d_coeff"** is incorrect and should be renamed to **"d_coef."**

**Details:**

- The parameter **"d_coeff"** does not affect training because its name is not recognized and the value is not passed to the trainer.
    
- In the [Prodigy repository](https://github.com/konstmish/prodigy), this parameter is correctly named **"d_coef."**
    
- I verified this by manually inputting `d_coef` in **extra_optimizer_args** (e.g., `d_coef=0.25` or `d_coef=2`). This successfully controlled the learning rate decay, resulting in significant changes in training output previews. In contrast, adjusting **"d_coeff"** in the panel had no effect, even when set to 0 or 2048.
    

This naming mistake prevents proper use of the Prodigy and Prodigy+ Schedule Free optimizer's most critical parameter and can mislead new users.

**Suggested Solution:**  
Rename the parameter **"d_coeff"** to **"d_coef"** to ensure correct functionality and consistency with the original optimizer implementation.

Thank you.

**Comment by kijai** [comment]
That's true, thanks, fixed. Also switched to using the pypi package as it's now available, so to keep using that optimizer the package should be installed:

`pip install prodigy-plus-schedule-free`

**Comment by Packsod** [comment]
Thank you for your prompt response. However, it appears that the `d_coef` parameter, despite being corrected in name, is still not being passed correctly into `optimizer_args=[]`. Instead, it is set outside, rendering it ineffective. As seen in the `args output info`, `d_coef=2048.0` is the output value from the Optimizer Config ProdigyPlusScheduleFree node panel (invalid), while `d_coef=0.05` is the manually entered parameter (valid).

For the same reason, Parameter `d0` is also invalid, adjusting it will not reflect any changes,

Since I'm not familiar with the training-related code and prefer not to make arbitrary changes, I'm unsure if other parameters for the Optimizer also need to be passed into `optimizer_args=[]`. Therefore, I leave this decision to your expertise.

> Namespace(console_log_level=None, console_log_file=None, console_log_simple=False, v2=False, v_parameterization=False, pretrained_model_name_or_path='Z:\\\\models_external\\\\comfyui_external_models\\\\checkpoints\\\\sdxl\\\\SocialCredit.safetensors', tokenizer_cache_dir=None, train_data_dir=None, cache_info=False, shuffle_caption=False, caption_separator=',', caption_extension='.caption', caption_extention=None, keep_tokens=0, keep_tokens_separator='', secondary_separator=None, enable_wildcard=False, caption_prefix=None, caption_suffix=None, color_aug=False, flip_aug=False, face_crop_aug_range=None, random_crop=False, debug_dataset=False, resolution=None, cache_latents=True, vae_batch_size=1, cache_latents_to_disk=True, skip_cache_check=False, enable_bucket=False, min_bucket_reso=256, max_bucket_reso=1024, bucket_reso_steps=64, bucket_no_upscale=False, token_warmup_min=1, token_warmup_step=0.0, alpha_mask=True, dataset_class=None, caption_dropout_rate=0.0, caption_dropout_every_n_epochs=0, caption_tag_dropout_rate=0.0, reg_data_dir=None, in_json=None, dataset_repeats=1, output_dir='workshop3:\\Project_Z_test002\\', output_name='pooh_1024_rank16_fp8_e4m3fn', huggingface_repo_id=None, huggingface_repo_type=None, huggingface_path_in_repo=None, huggingface_token=None, huggingface_repo_visibility=None, save_state_to_huggingface=False, resume_from_huggingface=False, async_upload=False, save_precision='fp8_e4m3fn', save_every_n_epochs=None, save_every_n_steps=None, save_n_epoch_ratio=None, save_last_n_epochs=None, save_last_n_epochs_state=None, save_last_n_steps=None, save_last_n_steps_state=None, save_state=False, save_state_on_train_end=False, resume=None, train_batch_size=1, max_token_length=None, mem_eff_attn=True, torch_compile=False, dynamo_backend='inductor', xformers=False, sdpa=False, vae=None, max_train_steps=800, max_train_epochs=None, max_data_loader_n_workers=0, persistent_data_loader_workers=False, seed=42, gradient_checkpointing=True, gradient_accumulation_steps=1, mixed_precision='bf16', full_fp16=False, full_bf16=False, fp8_base=False, fp8_dtype='e4m3', ddp_timeout=None, ddp_gradient_as_bucket_view=False, ddp_static_graph=False, clip_skip=2, logging_dir=None, log_with=None, log_prefix=None, log_tracker_name=None, wandb_run_name=None, log_tracker_config=None, wandb_api_key=None, log_config=False, noise_offset=None, noise_offset_random_strength=False, multires_noise_iterations=None, ip_noise_gamma=0.1, ip_noise_gamma_random_strength=False, multires_noise_discount=0.3, adaptive_noise_scale=None, zero_terminal_snr=False, min_timestep=None, max_timestep=None, loss_type='l2', huber_schedule='snr', huber_c=0.1, huber_scale=1.0, lowram=False, highvram=False, sample_every_n_steps=None, sample_at_first=False, sample_every_n_epochs=None, sample_prompts=\['Winnie the Pooh has nothing better to do, so he travels in space as an astronaut, looking cute yet silly.,'\], sample_sampler='ddim', config_file=None, output_config=False, metadata_title=None, metadata_author=None, metadata_description=None, metadata_license=None, metadata_tags=None, prior_loss_weight=1.0, conditioning_data_dir=None, masked_loss=False, deepspeed=False, zero_stage=2, offload_optimizer_device=None, offload_optimizer_nvme_path=None, offload_param_device=None, offload_param_nvme_path=None, zero3_init_flag=False, zero3_save_16bit_model=False, fp16_master_weights_and_gradients=False, optimizer_type='ProdigyPlusScheduleFree', use_8bit_adam=False, use_lion_optimizer=False, learning_rate=1.0000000000000004, max_grad_norm=1.0, optimizer_args=\['use_bias_correction=False', '**d_coef=0.05**'\], lr_scheduler_type='', lr_scheduler_args=None, lr_scheduler='constant', lr_warmup_steps=0, lr_decay_steps=0, lr_scheduler_num_cycles=1, lr_scheduler_power=1, fused_backward_pass=False, lr_scheduler_timescale=None, lr_scheduler_min_lr_ratio=None, dataset_config='\[\[datasets\]\]\\nresolution = \[ 1024, 1024,\]\\nbatch_size = 2\\nenable_bucket = true\\nbucket_no_upscale = false\\nmin_bucket_reso = 512\\nmax_bucket_reso = 1536\\n\[\[datasets.subsets\]\]\\nimage_dir = "Z:\\\\\\\\ProjectZ_test02\\\\\\\\caches\\\\\\\\comfyUI_cache\\\\\\\\lora_train\\\\\\\\tag_test\\\\\\\\"\\nclass_tokens = "pooh bear"\\nnum_repeats = 1\\n\\n\\n\[general\]\\nshuffle_caption = false\\ncaption_extension = ".txt"\\nkeep_tokens_separator = "|||"\\ncaption_dropout_rate = 0.0\\ncolor_aug = false\\nflip_aug = false\\n', cache_text_encoder_outputs=True, cache_text_encoder_outputs_to_disk=True, text_encoder_batch_size=None, disable_mmap_load_safetensors=False, weighting_scheme='uniform', logit_mean=0.0, logit_std=1.0, mode_scale=1.29, blocks_to_swap=0, min_snr_gamma=1.5000000000000002, scale_v_pred_loss_like_noise_pred=False, v_pred_like_loss=None, debiased_estimation_loss=False, weighted_captions=False, cpu_offload_checkpointing=False, no_metadata=False, save_model_as='safetensors', unet_lr=None, text_encoder_lr=None, fp8_base_unet=False, network_weights=None, network_module='.networks.lora', network_dim=16, network_alpha=1.0000000000000002, network_dropout=None, network_args=None, network_train_unet_only=True, network_train_text_encoder_only=False, training_comment=None, dim_from_weights=False, scale_weight_norms=None, base_weights=None, base_weights_multiplier=None, no_half_vae=False, skip_until_initial_step=False, initial_epoch=None, initial_step=None, negative_prompts=\['text, watermark, pig'\], num_cpu_threads_per_process=1, spda=True, lr=1.0000000000000004, prodigy_steps=40, d0=1.0000000000000004e-06, **d_coef=2048.0**, split_groups=True, use_stableadamw=True, use_cautious=False, use_adopt=False, use_grams=False, stochastic_rounding=True, use_orthograd=False, \*\*{'use_focus ': False})

**Comment by Packsod** [comment]
Alright, I’ve realized that it's better not to modify the **Optimizer Config ProdigyPlusScheduleFree** node further because:

- This node overlaps in functionality with the **Optimizer Config** node but lacks essential options. For example, **Prodigy** is typically used with a **cosine scheduler**, whereas the default output here is "constant," which cannot be changed via the panel. This leads to inevitable overfitting during the later stages of training.

Instead of modifying it, I suggest solely using the **Optimizer Config** node and appending additional options in the `extra_optimizer_args` input box.

As a side note, the `extra_optimizer_args` input box currently only allows passing a single argument and cannot correctly handle multiple formatted parameters. I will open a separate issue in detail.

Thank you for your efforts!
